### ISSUE TITLE
regression: ICE failed to get output type of async function

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#123273]
```
[INFO] [stdout] error: internal compiler error: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:1081:25: failed to get output type of async function
[INFO] [stdout]   --> src/components/rss_list.rs:84:40
[INFO] [stdout]    |
[INFO] [stdout] 84 |             .filter(move |&channel| -> Option<Channel> {
[INFO] [stdout]    |                                        ^^^^^^^^^^^^^^^
[INFO] [stdout] 
[INFO] [stdout] 
```

* https://crater-reports.s3.amazonaws.com/beta-1.78-3/beta-2024-03-30/gh/HiggRn.podcrab/log.txt

**Comment by matthiaskrgr** [comment]
could be related to https://github.com/rust-lang/rust/issues/105304 but may also be a red herring :fish:  :thinking:

**Comment by compiler-errors** [comment]
Regression in https://github.com/rust-lang-ci/rust/commit/cb682d38536a026e69702adaf24a23bfae6b4f04
The PR introducing the regression in this rollup is #120696: Properly handle `async` block and `async fn` in `if` exprs …

searched nightlies: from nightly-2024-01-01 to nightly-2024-03-18
regressed nightly: nightly-2024-02-14
searched commit range: https://github.com/rust-lang/rust/compare/b381d3ab27f788f990551100c4425bb782d26d76...a84bb95a1f65bfe25038f188763a18e096a86ab2
regressed commit: https://github.com/rust-lang/rust/commit/fd9bb7fddedddc2ab48665214efae780dc64af2d

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.8</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2024-01-01 --end=2024-03-18 --regress=ice 
```
</details>

**Comment by compiler-errors** [comment]
Seems like a quick diagnostics fix-forward. I'll put up a PR in a few hours when I'm done bisecting more PRs.

@rustbot claim

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23123273.20regression.3A.20ICE.20failed.20to.20get.20output.20type.20of.20async.20.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by cuviper** [comment]
#123804 was said to fix this, and I backported it in #124272, so I think this is done. It was called a minimal fix though, so if there's future work to be done, feel free to reopen or file a new issue.
